truffleruby/src/main/java/org/truffleruby/core/rubinius/IONodes.java

```diff
@@ -799,7 +801,9 @@ public int accept(DynamicObject io) {
             final int newFd;
 
             try {
-                newFd = ensureSuccessful(nativeSockets().accept(fd, memoryManager().newPointer(address), addressLength));
+                newFd = getContext().getThreadManager().runUntilResult(this,
+                        () -> ensureSuccessful(nativeSockets().accept(fd, memoryManager().newPointer(address), addressLength)),
+                        () -> nativeSockets().shutdown(fd, SHUT_RD));
```

Comment:
I don't know the intricacies of the call to say whether or not it permanently closes the FD. I should probably write a small C program that sees if `accept` can be called again. But if it can't, I'm not really sure what more we can do. As far as I can tell, MRI relies on `pthread_kill` with `SIGVTALRM` to interrupt the underlying thread.